rustls/src/msgs/message.rs

```diff
@@ -459,6 +459,149 @@ impl<'a> BorrowedPlainMessage<'a> {
     }
 }
 
+#[derive(Debug, Clone)]
+/// A collection of borrowed plaintext slices
+pub enum BorrowedPlainPayload<'a> {
+    Empty,
+    /// A single byte slice. This allows only one redirection instead of several.
+    Single(&'a [u8]),
+    /// A collection of chunks (byte slices)
+    /// and inclusive-exclusive cursors to single out a slice of bytes.
+    /// [BorrowedPayload] assumes three invariants:
+    /// - at least two chunks
+    /// - a start cursor pointing into the first chunk
+    /// - an end cursor pointing into the last one
+    Multiple {
+        chunks: &'a [&'a [u8]],
+        start: usize,
+        end: usize,
+    },
+}
+
+impl<'a> BorrowedPlainPayload<'a> {
+    /// Create a payload with a single byte slice
+    pub fn new_single(payload: &'a [u8]) -> Self {
+        Self::Single(payload)
+    }
+
+    /// create a payload from a slice of byte slices.
+    /// The cursors are added by default: start = 0, end = length
+    pub fn new(chunks: &'a [&'a [u8]]) -> Self {
+        Self::new_with_cursors(
+            chunks,
+            0,
+            chunks
+                .iter()
+                .map(|chunk| chunk.len())
+                .sum(),
+        )
+    }
+
+    /// Append all bytes to a vector
+    pub fn copy_to_vec(&self, vec: &mut Vec<u8>) {
+        match self {
+            Self::Empty => {}
+            Self::Single(chunk) => vec.extend_from_slice(chunk),
+            Self::Multiple { chunks, start, end } => {
+                let mut size = 0usize;
+                for (i, chunk) in chunks.iter().enumerate() {
+                    if i == 0 {
+                        vec.extend_from_slice(&chunk[*start..]);
+                        size += chunk.len();
+                    } else if i == chunks.len() - 1 {
+                        vec.extend_from_slice(&chunk[..(end - size)]);
+                    } else {
+                        vec.extend_from_slice(chunk);
+                        size += chunk.len();
+                    }
+                }
+            }
+        }
+    }
+
+    /// split self in two, around an index.
+    /// Works similarly to `split_at` in the core library.
+    pub fn split_at(&self, mid: usize) -> (Self, Self) {
+        match self {
+            Self::Empty => (Self::Empty, Self::Empty),
+            Self::Single(chunk) => {
+                if chunk.len() < mid {
+                    return (self.clone(), Self::Empty);
+                }
+                (Self::Single(&chunk[..mid]), Self::Single(&chunk[mid..]))
+            }
+            Self::Multiple { chunks, start, end } => {
+                let mut size = 0usize;
```

Comment:
I think we can use just `0` here - the compiler seems able to infer the `usize` type from use.